### PR TITLE
Support Vector Similarity

### LINKED ISSUE #146: Add dialect support for RediSearch queries
On RediSearch 2.4.3 we will introduce a new `DIALECT` command argument for queries (`FT.SEARCH`, `FT.AGGREGATE`) this will allow to process queries either by the current dialect (RediSearch 2.2.x) or the new dialect introduced on RediSearch 2.4.2.
New dialect features:
1. parameterized queries
2. VSS
3. Improved syntax behavior as introduced at https://github.com/RediSearch/RediSearch/releases/tag/v2.4.2

### PR DESCRIPTION
**Pull request by Avital-Fine**
closes #146

### CODE REVIEW

**Review by DvirDukhan — COMMENTED** [review]
Looks good
does the param `interface{}` type support binary blobs?

File: redisearch/query.go [hunk 1]
```diff
@@ -227,6 +229,18 @@ func (q Query) serialize() redis.Args {
...
+
+	if q.Params != nil {
+		args = args.Add("PARAMS", len(q.Params) * 2)
+		for name, value := range q.Params {
+			args = args.Add(name, value)
+		}
+	}
+
+	if q.Dialect != 0 {
+		args = args.Add("DIALECT", q.Dialect)
```
Comment (chayim) [inline redisearch/query.go:241]: Any reason to not add dialect anyways?

File: redisearch/query.go [hunk 2]
```diff
@@ -370,6 +384,29 @@ func (q *Query) SummarizeOptions(opts SummaryOptions) *Query {
...
+// e.g., $user , and each such reference in the search query to a parameter name is substituted
+// by the corresponding parameter value.
+func (q *Query) SetParams(params map[string]interface{}) *Query {
+	q.Params = params
+	return q
+}
+
+// AddParam adds a new param to the parameters list
+func (q *Query) AddParam(name string, value interface{}) *Query {
+	if q.Params == nil {
```
Comment (chayim) [inline redisearch/query.go:397]: Very good

**Review by chayim — APPROVED** [review]

**Comment by chayim** [comment]
@Avital-Fine the code looks right, but did you notice the CI failure? I think it's because we're using redisearch:1.6.15 and you probably want a more modern docker (redisfab?)

**Comment by Avital-Fine** [comment]
> @Avital-Fine the code looks right, but did you notice the CI failure? I think it's because we're using redisearch:1.6.15 and you probably want a more modern docker (redis-stack-server or redismod)

How do I update the docker then?
I was trying to add tests with the client now and it is failing

File: redisearch/redisearch_test.go [hunk 3]
```diff
@@ -510,3 +510,67 @@ func TestReturnFields(t *testing.T) {
 	assert.Equal(t, "Jon", docs[0].Properties["name"])
 	assert.Equal(t, "25", docs[0].Properties["years"])
 }
+
+func TestParams(t *testing.T) {
+	c := createClient("TestParams")
+	version, _ := c.getRediSearchVersion()
+	if version < 20200 {
```
Comment (Avital-Fine) [inline redisearch/redisearch_test.go:517]: @chayim Is it the right version?

File: redisearch/redisearch_test.go [hunk 4]
```diff
@@ -510,3 +510,67 @@ func TestReturnFields(t *testing.T) {
...
+		SetParams(map[string]interface{}{"min": "1", "max": "2"}).
+		SetDialect(2))
+	assert.Nil(t, err)
+	assert.Equal(t, 2, total)
+}
+
+func TestVectorField(t *testing.T) {
+	c := createClient("TestVectorField")
+	version, _ := c.getRediSearchVersion()
+	if version < 20200 {
```
[the remark this answers is not in the document]
Comment (chayim) [reply redisearch/redisearch_test.go:544]: No, VecSim is ```version > 20430``` I think, 2.4.3.

File: redisearch/redisearch_test.go (hunk 3 above)
Comment (chayim) [reply redisearch/redisearch_test.go:517]: Same as above

**Review by chayim — APPROVED** [review]

**Pull request merged**